cob_script_server/src/simple_script_server/simple_script_server.py

```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
 		point_nr = 0
 		traj_time = 0
 
-		param_string = self.ns_global_prefix + "/" + component_name + "/default_vel"
-		if not rospy.has_param(param_string):
-			default_vel = 0.1 # rad/s
-			rospy.logwarn("parameter %s does not exist on ROS Parameter Server, using default of %f [rad/sec].",param_string,default_vel)
-		else:
-			default_vel = rospy.get_param(param_string)
+		if default_vel: # passed via argument
+			rospy.logdebug("using default_vel from argument")
+			if (type(default_vel) is float) or (type(default_vel) is int):
+				default_vel = numpy.array([default_vel for _ in start_pos])
+			elif (type(default_vel) is list) and (len(default_vel) == len(start_pos)) and all(((type(item) is float) or (type(item) is int)) for item in default_vel):
+				default_vel = default_vel
+			else:
+				rospy.logerr("parameter %s has wrong format (must be float/int or list of float/int) with proper dimensions.")
+				return (JointTrajectory(), 3)
+		else: # get from parameter server
+			rospy.logdebug("using default_vel parameter server")
+			param_string = self.ns_global_prefix + "/" + component_name + "/default_vel"
+			if not rospy.has_param(param_string):
+				default_vel = numpy.array([0.1 for _ in start_pos]) # rad/s
+				rospy.logwarn("parameter %s does not exist on ROS Parameter Server, using default_vel {} [rad/sec].".format(param_string,default_vel))
+			else:
+				param_vel = rospy.get_param(param_string)
+				if (type(param_vel) is float) or (type(param_vel) is int):
+					default_vel = numpy.array([param_vel for _ in start_pos])
+				elif (type(param_vel) is list) and (len(param_vel) == len(start_pos)) and all(((type(item) is float) or (type(item) is int)) for item in param_vel):
+					default_vel = param_vel
+				else:
+					default_vel = numpy.array([0.1 for _ in start_pos]) # rad/s
+					rospy.logwarn("parameter %s has wrong format (must be float/int or list of float/int), using default_vel {} [rad/sec].".format(param_string,default_vel))
+		rospy.logdebug("default_vel: {}".format(default_vel))
+
+		robot_urdf = URDF.from_parameter_server()
+		limit_vel = []
+		for idx, joint_name in enumerate(joint_names):
+			try:
+				limit_vel.append(robot_urdf.joint_map[joint_name].limit.velocity)
+			except KeyError:
+				limit_vel.append(default_vel[i])
 
+		# limit_vel from urdf or default_vel (from argument or parameter server)
 		if urdf_vel:
-			robot_urdf = URDF.from_parameter_server()
-			velocities = []
-			for joint_name in joint_names:
-				try:
-					velocities.append(robot_urdf.joint_map[joint_name].limit.velocity)
-				except KeyError:
-					velocities.append(default_vel)
+			rospy.logdebug("using default_vel from urdf_limits")
+			velocities = limit_vel
 		else:
-			velocities = [default_vel for _ in joint_names]
-		rospy.loginfo("Velocities are: {}".format(velocities))
+			rospy.logdebug("using default_vel from argument or parameter server")
+			velocities = list(default_vel)
```

Comment:
Why convert back to list here?

Reply:
this way I did not have to re-write what's coming below...:wink: